[vulkan] inplace add_, relu_

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42181 [vulkan] Permtute ops: transpose,view,slice,unsqueeze,select
* #42180 [vulkan][op] avg_pool2d
* #42179 [vulkan][op] add.Scalar, mul.Scalar
* #41434 [vulkan] cat op (concatenate)
* **#41380 [vulkan] inplace add_, relu_**

Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

IvanKobzarev committed:
[vulkan] inplace add_, relu_

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -272,6 +272,18 @@ Tensor vulkan_add(const Tensor& self, const Tensor& other, const Scalar alpha) {
 
+Tensor& vulkan_add_(Tensor& self, const Tensor& other, Scalar alpha) {
+  auto& x = vtensor_from_vulkan(self.is_vulkan() ? self : self.vulkan());
+  auto& y = vtensor_from_vulkan(other.is_vulkan() ? other : other.vulkan());
```
const.  I think vtensor_from_vulkan should be overloaded on const.

IvanKobzarev replied:
I will add const override for vtensor_from_vulkan to this PR

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -272,6 +272,18 @@ Tensor vulkan_add(const Tensor& self, const Tensor& other, const Scalar alpha) {
+  output.allocate_storage();
+  vulkan::detail::add(output, x, y, a);
+  x = std::move(output);
+  return self;
```
I'm not sure I understand this logic here.  If self is already allocated and the goal of this function is to implement c += a * b, then do we need to allocate output as well?  Can we not do it in place in 'self'?  Also, why the move?  Can't we just pass output, output, and input and a (in that order) to vulkan::detail::add?

IvanKobzarev replied:
Here we read from x (self) and y (other) and write output values to output.  
At the moment all shaders work having separate output and input(s), here it is using the same shader that is used for copy variant of this op.

`x = std::move(output)` replaces x (self) content (pointer to VulkanTensor::Impl) of x to the result of output.

`vulkan::detail::add()` uses only `image()` representation, so we do not really need allocated buffer(). 
For this implementation we need `output.allocate_storage()`, but with changes  https://github.com/pytorch/pytorch/pull/42569 allocate_storage can be removed

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -413,23 +425,30 @@ Tensor vulkan_clamp(
+      x,
+      min ? min.value().to<float>() : -std::numeric_limits<float>::infinity(),
+      max ? max.value().to<float>() : std::numeric_limits<float>::infinity());
+  x = std::move(output);
```
Same.  If this operation is in place, then why do we need to `allocate_storage`?

IvanKobzarev replied:
(As in previous comment) Output VulkanTensor is used here as a temporary variable to reuse shaders with separate output, inputs; allocate_storage to remove in https://github.com/pytorch/pytorch/pull/42569 :)

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] inplace add_, relu_"


Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

[ghstack-poisoned]
